### DISCUSSION TITLE
Extending the -x syntax

### DISCUSSION THREAD

**Discussion by pbsds** [sharkdp/fd#1135] [Ideas]
I love `fd`, and use it almost daily. Especially do i love the easy-to-use `-x` syntax.
But i do often find myself doing tricks like `fd ... -x sh -c 'foobar "$(basename "{//}")"'` to get the basename of the parent directory.

I decided to have a look at the `-x` syntax, and I propose this extension:

```
The following placeholders are substituted before the command is executed:

     {}     path (of the current search result)

     {/}    basename (deprecated)

     {-}    basename

     {//}   parent directory

     {///}  parent parent directory (repeatable)

     {//-}  parent directory basename

     {.}    path without outermost file extension

     {..}   path without the two outermost file extensions (repeatable)

     {/.}   basename without file extension (deprecated)

     {-.}   basename without file extension
```

The changes are:

* Change basename from `/` to `-`, as it conflicts with parent name, deprecating the old syntax
* Enable repeated applications of `/` and `.` (related: #911)

This will change the general find-replace regex match from `\{/*\.*\}` to `\{/*-?.*\}`.

I kept two deprecations for backward-compatibility, as people most likely use `fd` in in scripts. After a while, the documentation of the deprecated syntax could be removed, but we could keep their implementation and yield a warning.

Also something we could discuss: Should it be possible to escape the syntax? Something like
`fd . -x echo "{}\{}"` could print `myfile.txt{}`

**Comment by sharkdp** [comment]
Thank you for writing this up. Our syntax is based on that of GNU parallel (https://www.gnu.org/software/parallel/), so I'd rather not introduce any breaking changes, because that would also mean we're not compatible with `parallel` anymore.

But maybe we can study their documentation/issue tracker to see if there are extensions to the syntax we have in `fd`?

**Comment by nyurik** [comment]
A common pattern i run into is the ability to do search/replace in a path.  For example, if I have many test fixtures in `../test/fixtures/.../*.pbf` and i need to convert them all to `../test/expected/.../*.mvt`, I have to do something weird like this:

```
fd . ../test/fixtures -e pbf -x echo my_converter  {}   $(tmp='{.}'; echo ${tmp/\/fixtures\///expected/}).mvt
```


It might be better to have a regex search/replace pattern right in the `{...}` syntax if possible.  I do not think support for parallel is that critical tbh, esp if it does not break existing parallel features, but rather extends them

example for the above:   `$(tmp='{.}'; echo ${tmp/\/fixtures\///expected/})` can be expressed as `{.|/\/fixtures\///expected/}` -- basically regex is on after the first `/` and until the next `/` that hasn't been escaped -- `/fixtures/` -> `/expected/`

**Reply by pbsds** [reply]
You may be interested in https://github.com/bulletmark/edir, where you can find-replace using a text editor and double-check the results before committing. Alternatives include https://github.com/thameera/vimv and its rust variant https://www.dmulholl.com/dev/vimv.html

**Reply by nyurik** [reply]
@pbsds thx, but my point was not about modifying the file -- for that I have my own custom file processor. I just needed to run my command for each test file I have, and generate corresponding file in the `expected/` dir - a common paradigm for the integration testing of my code.
